File: src/main/java/com/tambapps/http/restclient/request/body/AbstractBodyProcessor.java
package com.tambapps.http.restclient.request.body;

import java.io.IOException;
import java.io.OutputStream;
import java.net.URLConnection;

/**
 * Class to extend in order to implement a custom {@link BodyProcessor}
 */
public abstract class AbstractBodyProcessor implements BodyProcessor {

  @Override
  public final void prepareConnection(URLConnection connection) throws IOException {
    connection.setDoOutput(true);
    prepareURLConnection(connection);
    try (OutputStream oStream =  connection.getOutputStream()) {
      writeContent(oStream);
    }
  }

  protected abstract void writeContent(OutputStream oStream) throws IOException;

  protected void prepareURLConnection(URLConnection connection) {}
}

File: httpclient-core/src/main/kotlin/com/tambapps/http/client/request/body/AbstractBodyProcessor.kt
package com.tambapps.http.client.request.body

import java.io.IOException
import java.io.OutputStream
import java.net.URLConnection

/**
 * Class to extend in order to implement a custom [BodyProcessor]
 */
abstract class AbstractBodyProcessor : BodyProcessor {
    @Throws(IOException::class)
    override fun prepareConnection(connection: URLConnection) {
        connection.doOutput = true
        prepareURLConnection(connection)
        connection.getOutputStream().use { oStream -> writeContent(oStream) }
    }

    @Throws(IOException::class)
    protected abstract fun writeContent(oStream: OutputStream)
    protected open fun prepareURLConnection(connection: URLConnection) {}
}